### PR TITLE
Replace magic hasReturnExp bitflags with fields

### PR DESCRIPTION
**Pull request by dkorpel**
The throw statement and assert(0) ones seem unused.

### CODE REVIEW

File: compiler/src/dmd/expressionsem.d [hunk 1]
```diff
@@ -8135,8 +8135,6 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             /* This is an `assert(0)` which means halt program execution
              */
             FuncDeclaration fd = sc.parent.isFuncDeclaration();
-            if (fd)
-                fd.hasReturnExp |= 4;
```
Comment (thewilsonator) [inline compiler/src/dmd/expressionsem.d:8139]: why is this removed with no replacement?

File: compiler/src/dmd/statementsem.d [hunk 2]
```diff
@@ -3686,9 +3687,6 @@ public bool throwSemantic(const ref Loc loc, ref Expression exp, Scope* sc)
         return false;
     }
 
-    if (FuncDeclaration fd = sc.parent.isFuncDeclaration())
-        fd.hasReturnExp |= 2;
-
```
Comment (thewilsonator) [inline compiler/src/dmd/statementsem.d:3691]: ditto

**Review by thewilsonator — APPROVED** [review]
This otherwise looks great! 

Silly questions:

**Comment by thewilsonator** [comment]
Oh I see you said that in the opening comment.

**Comment by thewilsonator** [comment]
Hmm. I still think it would be good to leave those in there, DMDaaL might want that. @RazvanN7 ?

**Comment by RazvanN7** [comment]
It has happened in the past that fields that were unused in dmd were actually being used by users of dmdlib. That being said, I'm not sure what the policy should be here. Normally, you should be able to modify the compiler code however you wish, but this means breaking people's code. To a certain extent that's a given (and not a problem, provided that there are alternative ways in which you can obtain the needed information), however, there are situations where the compiler stores some information that is very hard to obtain if one would not have the ability to simply access it (see [1] for an example and  its revert [2]).

As a compiler developer, I understand the need to optimize the code as much as possible. As a dmdlib library user I'd wish that the compiler stores as much information as possible so that I can use it. I think that the best way to proceed, generally, is to not delete information that *might* be useful, even if at a first glance it seems that the compiler is not using it. In this particular case, I will have Vladut I have look - it might be possible that we are using some of the deleted bits in D-scanner.

[1] https://github.com/dlang/dmd/pull/15241
[2] https://github.com/dlang/dmd/pull/16193

**Review by RazvanN7 — APPROVED** [review]
